File: pkg/models/query.go
package models

type QueryType string

const (
	QueryCollections     = "collections"
	QueryFind            = "find"
	QueryCount           = "count"
	QueryAggregate       = "aggregate"
	QueryDatabaseStats   = "databasestats"
	QueryCollectionStats = "collectionstats"
)

type QueryModelFind struct {
	Collection string `json:"collection"`
	Filter     string `json:"filter"`
	Sort       string `json:"sort"`
	Limit      int64  `json:"limit"`
	Explain    bool   `json:"explain"`
	Verbosity  string `json:"verbosity"`
}

type QueryModelCount struct {
	Collection string `json:"collection"`
	Filter     string `json:"filter"`
}

type QueryModelAggregate struct {
	Collection string `json:"collection"`
	Pipeline   string `json:"pipeline"`
	Explain    bool   `json:"explain"`
	Verbosity  string `json:"verbosity"`
}

type QueryModelCollectionStats struct {
	Collection          string `json:"collection"`
	IncludeIndexDetails bool   `json:"includeIndexDetails"`
	IncludeWiredTiger   bool   `json:"includeWiredTiger"`
}

File: src/query.ts
import { Query, QueryType } from './types';

// Fields that must be set for a query of a given type to be executed. This is
// the single source of truth shared by DataSource.filterQuery (which decides
// whether a query is run) and the query editor (which warns the user about
// missing fields), so the two never drift apart.
export const REQUIRED_FIELDS: Record<QueryType, Array<keyof Query>> = {
  collections: [],
  find: ['collection', 'filter', 'sort', 'limit'],
  count: ['collection', 'filter'],
  aggregate: ['collection', 'pipeline'],
  databasestats: [],
  collectionstats: ['collection'],
  indexes: [],
  indexstats: ['collection', 'index'],
};

const FIELD_LABELS: Partial<Record<keyof Query, string>> = {
  collection: 'Collection',
  filter: 'Filter',
  sort: 'Sort',
  limit: 'Limit',
  pipeline: 'Pipeline',
  index: 'Index',
};

export const getMissingRequiredFields = (query: Query): string[] => {
  const required = REQUIRED_FIELDS[query.queryType] ?? [];
  return required
    .filter((field) => !query[field])
    .map((field) => FIELD_LABELS[field] ?? field);
};
